src/spatial/nsm.jl

```diff
@@ -0,0 +1,257 @@
+        for site in [source_site, target_site]
+            update_site_time!(p.pq, p.rng, p.cur_rates, site, t)
+        end
+    else
+        site = jump.src
+        update_rates_after_reaction!(p, u, site, reaction_id_from_jump(p,jump))
+        update_site_time!(p.pq, p.rng, p.cur_rates, site, t)
+    end
+end
+
+function update_site_time!(pq, rng, cur_rates, site, t)
+    site_rate = total_site_rate(cur_rates, site)
+    if site_rate > zero(typeof(site_rate))
+        update!(pq, site, t + randexp(rng) / site_rate)
+    else
+        update!(pq, site, typemax(t))
+    end
+end
+
+######################## helper routines for all spatial SSAs ########################
+function update_rates_after_reaction!(p, u, site, reaction_id)
+    update_reaction_rates!(p.cur_rates, p.dep_gr[reaction_id], u, p.ma_jumps, site)
+    update_diffusion_rates!(p.cur_rates, p.jumptovars_map[reaction_id], p.diffusion_constants, u, site, p.spatial_system)
+end
+
+function update_rates_after_diffusion!(p, u, source_site, target_site, species)
+    update_reaction_rates!(p.cur_rates, p.vartojumps_map[species], u, p.ma_jumps, source_site)
+    update_diffusion_rates!(p.cur_rates, species, p.diffusion_constants, u, source_site, p.spatial_system)
+    
+    update_reaction_rates!(p.cur_rates, p.vartojumps_map[species], u, p.ma_jumps, target_site)
+    update_diffusion_rates!(p.cur_rates, species, p.diffusion_constants, u, target_site, p.spatial_system)
+end
+
+"""
+update rates of all reactions in rxs at site
+"""
+function update_reaction_rates!(cur_rates, rxs, u, ma_jumps, site)
+    for rx in rxs
+        set_rx_rate_at_site!(cur_rates, site, rx, evalrxrate((@view u[:,site]), rx, ma_jumps))
+    end
+end
+
+"""
+update rates of all specs in list species at site
+"""
+function update_diffusion_rates!(cur_rates, species::AbstractArray, diffusion_constants, u, site, spatial_system)
+    for spec in species
+        set_hop_rate_at_site!(cur_rates, site, spec, evaldiffrate(diffusion_constants, u, spec, site, spatial_system))
+    end
+end
+"""
+update rates of species at site
+"""
+function update_diffusion_rates!(cur_rates, species, diffusion_constants, u, site, spatial_system)
+    set_hop_rate_at_site!(cur_rates, site, species, evaldiffrate(diffusion_constants, u, species, site, spatial_system))
+end
+
+"""
+update_state!(p, integrator)
+
+updates state based on p.next_jump
+"""
+function update_state!(p, integrator)
+    jump = p.next_jump
+    if is_diffusion(p, jump)
+        execute_diffusion!(integrator, jump.src, jump.dst, jump.jidx)
+    else
+        rx_index = reaction_id_from_jump(p,jump)
+        executerx!((@view integrator.u[:,jump.src]), rx_index, p.ma_jumps)
+    end
+    # save jump that was just exectued
+    p.prev_jump = jump
+    nothing
+end
+
+"""
+    is_diffusion(p, jump)
+
+true if jump is a diffusion
+"""
+function is_diffusion(p, jump)
+    jump.jidx <= size(p.diffusion_constants,1)
```

Comment:
I wonder if we should just store a `numspecies` field in `p` and calculate this in the constructor? This might make a bunch of the code cleaner.

Reply:
We can leave this one for future cleanup.